core/arch/riscv64/mangle.c

```diff
@@ -385,13 +394,406 @@ mangle_rel_addr(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+    PRE(ilist, next_instr,
+        instr_create_save_to_tls(dcontext, scratch_reg1, TLS_LRSC_SIZE_SLOT));
+
+    /* Restore scratch register(s). */
+    PRE(ilist, next_instr, instr_create_restore_from_tls(dcontext, scratch_reg1, slot1));
+    if (uses_reg_stolen) {
+        PRE(ilist, next_instr,
+            instr_create_save_to_tls(dcontext, scratch_reg2, TLS_REG_STOLEN_SLOT));
+        PRE(ilist, next_instr,
+            instr_create_restore_from_tls(dcontext, scratch_reg2, slot2));
+    }
+    return next_instr;
+}
+
+static instr_t *
+mangle_exclusive_store(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                       instr_t *next_instr)
+{
+    reg_id_t scratch_reg1, scratch_reg2;
+    ushort slot1, slot2;
+    int opcode;
+    opnd_t dst, src0;
+    opnd_size_t opsz;
+    instr_t *fail = INSTR_CREATE_label(dcontext), *final = INSTR_CREATE_label(dcontext),
+            *loop = INSTR_CREATE_label(dcontext), *end = INSTR_CREATE_label(dcontext);
+    ASSERT(instr_is_exclusive_store(instr));
+    ASSERT(instr_num_dsts(instr) == 1 && instr_num_srcs(instr) == 3);
+
+    /* TODO i#3544: Not implemented.  */
+    ASSERT_NOT_IMPLEMENTED(!instr_uses_reg(instr, dr_reg_stolen));
+    ASSERT_NOT_IMPLEMENTED(!instr_uses_reg(instr, DR_REG_TP));
+
+    dst = instr_get_dst(instr, 0);
+    src0 = instr_get_src(instr, 0);
+    opsz = instr_get_opcode(instr) == OP_sc_d ? OPSZ_8 : OPSZ_4;
+    scratch_reg1 = pick_scratch_reg(dcontext, instr, DR_REG_NULL, &slot1);
+    scratch_reg2 = pick_scratch_reg(dcontext, instr, scratch_reg1, &slot2);
+
+    /* Spill scratch registers. */
+    PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg1, slot1));
+    PRE(ilist, instr, instr_create_save_to_tls(dcontext, scratch_reg2, slot2));
+
+    /* Restore address saved by exclusive load and check if it's equal to src0. */
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_ADDR_SLOT));
+    PRE(ilist, instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(fail), opnd_create_reg(scratch_reg1),
+                         src0));
+
+    /* Restore size saved by exclusive load and check if it's equal to current size. */
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_SIZE_SLOT));
+    PRE(ilist, instr,
+        XINST_CREATE_load_int(dcontext, opnd_create_reg(scratch_reg2),
+                              opnd_add_flags(opnd_create_immed_int(opsz, OPSZ_12b),
+                                             DR_OPND_IMM_PRINT_DECIMAL)));
+    PRE(ilist, instr,
+        INSTR_CREATE_bne(dcontext, opnd_create_instr(fail), opnd_create_reg(scratch_reg1),
+                         opnd_create_reg(scratch_reg2)));
+
+    PRE(ilist, instr,
+        instr_create_restore_from_tls(dcontext, scratch_reg1, TLS_LRSC_VALUE_SLOT));
+    PRE(ilist, instr, loop);
+
+    /* Convert exclusive store to a compare-and-swap.
+     * Begin of the LR/SC sequence.
```

Comment:
Please add a comment somewhere showing a full sample code expansion to help the reader understand this.  For arm we have ldstex.dox which discusses the design and gives examples.